apis/core/v1beta1/openstackcontrolplane_webhook.go

```diff
@@ -46,14 +49,42 @@ var _ webhook.Validator = &OpenStackControlPlane{}
 func (r *OpenStackControlPlane) ValidateCreate() error {
 	openstackcontrolplanelog.Info("validate create", "name", r.Name)
 
-	return r.ValidateServices()
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, r.ValidateServices()...)
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{
+				Group: "openstackcontrolplane.openstack.org",
+				Kind:  "OpenstackControlplane",
+			},
+			r.Name, allErrs)
+	}
+
+	return nil
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
 func (r *OpenStackControlPlane) ValidateUpdate(old runtime.Object) error {
 	openstackcontrolplanelog.Info("validate update", "name", r.Name)
 
-	return r.ValidateServices()
+	var allErrs field.ErrorList
+
+	if err := r.ValidateServices(); err != nil {
+		allErrs = append(allErrs, err...)
+	}
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{
+				Group: "openstackcontrolplane.openstack.org",
+				Kind:  "OpenstackControlplane",
+			},
+			r.Name, allErrs)
+	}
```

Comment:
Isn't this the same code for both ValidateCreate() and ValidateUpdate() and can be in ValdateServices()? We don't use old object for update.

Reply:
I imagine we will run other validations than `ValidateServices`, each validation function returns a list of errors and append. We can change it to `ValidateServices()` and then have it call a `ValidateServiceDependencies()` and add other validations in `ValidateServices()` as they get introduced. ?

Reply:
OK, I was thinking that we would probably not do any validation specific for update/patch and compare with old object. But possibly there could be a few like changing the ml2 driver etc.